journalpump/senders/websocket.py

```diff
@@ -0,0 +1,286 @@
+from .base import LogSender
+from aiohttp_socks import ProxyConnectionError, ProxyError, ProxyTimeoutError
+from aiohttp_socks.utils import Proxy
+from concurrent.futures import TimeoutError as ConnectionTimeoutError
+from janus import Queue
+from threading import Thread
+from urllib.parse import urlparse
+
+import asyncio
+import logging
+import random
+import socket
+import ssl
+import time
+import websockets
+
+
+class ExponentialBackoff:
+    """
+    An implementation of the exponential backoff algorithm with full jitter described at
+    https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+    """
+
+    def __init__(self, base: float, factor: float, maximum: float, jitter: bool):
+        if base <= 0:
+            raise ValueError("base must be positive")
+        if factor <= 0:
+            raise ValueError("factor must be positive")
+        if maximum < base:
+            raise ValueError("maximum must be greater than or equal to base")
+        self._base = base
+        self._factor = factor
+        self._maximum = maximum
+        self._jitter = jitter
+        self._attempts = 0
+
+    def next_sleep(self) -> float:
+        result = self._base * (self._factor ** self._attempts)
+        if result <= self._maximum:
+            self._attempts += 1
+        else:
+            result = self._maximum
+
+        if self._jitter:
+            return random.uniform(0, result)
+
+        return result
+
+    def reset(self) -> None:
+        self._attempts = 0
+
+
+class WebsocketRunner(Thread):
+    def __init__(self, *, websocket_uri, socks5_proxy_url, ssl_enabled, ssl_ca, ssl_key, ssl_cert):
+        super().__init__()
+        self.setDaemon(True)
+        self.log = logging.getLogger(self.__class__.__name__)
+        self.websocket_uri = websocket_uri
+        self.socks5_proxy_url = socks5_proxy_url
+        self.ssl_enabled = ssl_enabled
+        self.ssl_ca = ssl_ca
+        self.ssl_key = ssl_key
+        self.ssl_cert = ssl_cert
+        self.websocket_loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(self.websocket_loop)
+        self.stop_event = asyncio.Event()
+        self.queue = None
+        self.running = True
+        self.connected = False
+        self.backoff = ExponentialBackoff(
+            base=10, factor=1.8, maximum=60, jitter=True
+        )
+        # prevent websockets from logging message contents when we're otherwise in DEBUG mode
+        logging.getLogger("websockets").setLevel(logging.INFO)
+
+        self.socks5_proxy = None
+        if self.socks5_proxy_url:
+            self.socks5_proxy = Proxy.from_url(self.socks5_proxy_url, loop=self.websocket_loop)
+
+    async def consumer_handler(self, websocket):
```

Comment:
Journalpump at the moment doesn't have any async code. Did you check if there are any good quality sync websocket clients that could be used instead? Since the same model than what you have here is successfully used elsewhere too I don't think it's a problem as such but just seems like synchronous client might have been more suitable in journalpump context. Now the async handling just adds some inefficiences when switching contexts and makes the code more complex.

Reply:
There's one which meets the basic requirements (active project, SOCKS5 support) at https://github.com/websocket-client/websocket-client but it does run another thread as well to manage the connection, so the context switching will still be around. Also has a bug from 2019 where it gets stuck and does not recover after a network connectivity failure (https://github.com/websocket-client/websocket-client/issues/592), although there might be a fix found for it now. An old version is packaged in Fedora, we'd need to package it ourself (and likely a couple of dependencies). Would feel safer to go with the already-familiar client at this point and reconsider later if necessary.